### PR TITLE
ARROW-7739: [GLib] Use placement new to initialize shared_ptr object in private structs

### PR DESCRIPTION
**Pull request by mrkn**
I inserted placement new for all shared pointer fields.

### CODE REVIEW

**Comment by pitrou** [comment]
cc @kou

**Comment by wesm** [comment]
Seems like a good idea.

File: c_glib/arrow-glib/array-builder.cpp [hunk 1]
```diff
@@ -4301,12 +4289,9 @@ G_DEFINE_TYPE_WITH_PRIVATE(GArrowStructArrayBuilder,
...
 garrow_struct_array_builder_dispose(GObject *object)
 {
-  GArrowStructArrayBuilderPrivate *priv;
-  GList *node;
-
-  priv = GARROW_STRUCT_ARRAY_BUILDER_GET_PRIVATE(object);
+  auto priv = GARROW_STRUCT_ARRAY_BUILDER_GET_PRIVATE(object);
 
-  for (node = priv->field_builders; node; node = g_list_next(node)) {
+  for (GList *node = priv->field_builders; node; node = g_list_next(node)) {
```
Comment (kou) [inline c_glib/arrow-glib/array-builder.cpp]: Can we use auto here?

**Review by kou — CHANGES_REQUESTED** [review]
Could you also call destructor explicitly in `finalize()`?

**Review by kou — APPROVED** [review]
+1

**Pull request closed without merging**